pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go

```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
 func newReconcileError(reason string, err error) *reconcileError {
 	return &reconcileError{degradationReason: reason, err: err}
 }
+
+// generatePublicKey generates a public key and returns it in the ssh-rsa..
+// string format using the retrieved private key
+func generatePublicKey() (string, error) {
+	privateKey, err := retrievePrivateKey()
+	if err != nil {
+		return "", fmt.Errorf("failed to retrieve private key in rsa format: %v", err)
+	}
+	publicKey, err := ssh.NewPublicKey(&privateKey.PublicKey)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate public key: %v", err)
+	}
+	pubKeyBytes := ssh.MarshalAuthorizedKey(publicKey)
+	return string(pubKeyBytes[:]), nil
+}
+
+// retrievePrivateKey retrieves private key in format required for generating
+// public key using the privateKeyPath
+func retrievePrivateKey() (*rsa.PrivateKey, error) {
+	privateKeyBytes, err := ioutil.ReadFile(wkl.PrivateKeyPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find private key from path: %v with: %v", wkl.PrivateKeyPath, err)
+	}
+
+	privateKeyBlock, _ := pem.Decode(privateKeyBytes)
+	if privateKeyBlock == nil {
+		return nil, fmt.Errorf("failed to decode retrieved private key: %v", wkl.PrivateKeyPath)
+	}
+
+	var parsedKey interface{}
+	if parsedKey, err = x509.ParsePKCS1PrivateKey(privateKeyBlock.Bytes); err != nil {
+		return nil, fmt.Errorf("failed to parse private key with: %v", err)
+	}
+
+	privateKey, ok := parsedKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, fmt.Errorf("failed to cast private key as rsa key type: %v", err)
+	}
+	return privateKey, nil
+}
+
+// createUserDataSecret creates a secret in 'openshift-machine-api' namespace
+// This secret will be used to inject cloud provider user data for creating
+// windows machines
+func (r *ReconcileWindowsMachineConfig) createUserDataSecret() error {
+
+	userDataSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "windows-user-data",
+			Namespace: "openshift-machine-api",
+		},
+		Data: map[string][]byte{
+			"userData": []byte(`<powershell>
+	$url = "https://raw.githubusercontent.com/ansible/ansible/devel/examples/scripts/ConfigureRemotingForAnsible.ps1"
+	$file = "$env:temp\ConfigureRemotingForAnsible.ps1"
+	(New-Object -TypeName System.Net.WebClient).DownloadFile($url,  $file)
+	& $file
+	Add-WindowsCapability -Online -Name OpenSSH.Server~~~~0.0.1.0
+	New-NetFirewallRule -DisplayName "` + types.FirewallRuleName + `"
```

Comment:
nit: I'd just define the FirewallRuleName and ContainerLogsPort here, 2 more vars which are not dependent on WNI's types.